### PR TITLE
Accept & CBOR support

### PR DESCRIPTION
**Pull request by Half-Shot**
This adds support for CBOR decoding and encoding in synapse, as well as `Accept` support which is part of matrix-org/matrix-doc#1740. **Please do not merge or review this just yet**

### CODE REVIEW

File: synapse/http/server.py [hunk 1]
```diff
@@ -415,24 +415,59 @@ def respond_with_json(request, code, json_object, send_cors=False,
...
         response_code_message=response_code_message,
+        encoding=encoding
     )
 
+def select_content_type(request):
+    accepts = request.getHeader(b"Accept")
+    if accepts is None:
+        return "application/json"
+    mimetypes = accepts.decode().split(",")
+    mimetypes = [mt.strip().lower().split(";q=")for mt in mimetypes]
```
Comment (Half-Shot) [inline synapse/http/server.py]: Doesn't handle `*/*` or `sometype/*`

File: synapse/http/servlet.py [hunk 2]
```diff
@@ -198,6 +199,17 @@ def parse_json_value_from_request(request, allow_empty_body=False):
     if not content_bytes and allow_empty_body:
         return None
 
+    contentType = request.getHeader(b"Content-Type")
```
Comment (Half-Shot) [inline synapse/http/servlet.py]: If content type is not supported we should 415 here.

File: synapse/http/server.py [hunk 3]
```diff
@@ -415,24 +415,59 @@ def respond_with_json(request, code, json_object, send_cors=False,
...
-            json_bytes = json.dumps(json_object).encode("utf-8")
+            # Check what encoding the client prefers.
+            encoding = select_content_type(request)
+            logger.debug(
+                "Selected encoding %s.",
+                encoding)
+            if encoding == "application/cbor":
+                json_bytes = cbor.dumps(json_object)
+            else: # Default to json
+                logger.warn(
```
Comment (Half-Shot) [inline synapse/http/server.py]: We should throw a 406 instead.

File: synapse/http/server.py (hunk 1 above)
Comment (hawkowl) [reply synapse/http/server.py]: This won't work on Python 3. (as an aside, do you want some help with unit tests for this code?)

Comment (Half-Shot) [reply synapse/http/server.py]: It's currently going through spec review land and this is just a demo, but yeah seeing as it's likely to be fine I'll probably like a hand with it :)

**Comment by hawkowl** [comment]
I'm going to close this until/if the MSC is accepted.

**Comment by hawkowl** [comment]
(just to clean up the PR list of things that aren't actionable)

**Pull request closed without merging**